master/buildbot/worker/base.py

```diff
@@ -781,6 +781,8 @@ def buildFinished(self, sb):
         if not self.building:
             if self.build_wait_timeout == 0:
                 d = self.insubstantiate()
+                d.addCallback(lambda _:
+                              AbstractWorker.disconnect(self))
```

Comment:
> This change makes sense to me to be added to insubstantiate, [...]

It might. But I'd be inclined to reject any change to this code that is "I think this makes sense ..." instead of "This makes sense because a) ... b) ... c) ...". I suspect the code is in the state that it is largely because of changes like the former.